File: Tree/Hard/145-Binary-Tree-Postorder-Traversal/solution_recursive.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public IList<int> PostorderTraversal(TreeNode root) {
        // recursive
        // tc:O(n); sc:O(h)
        if (root == null) {
            return new List<int>();
        }
        
        var left = PostorderTraversal(root.left) as List<int>;
        var right = PostorderTraversal(root.right);        
        
        left.AddRange(right);
        left.Add(root.val);
        
        return left;
    }
}

File: Tree/Hard/145-Binary-Tree-Postorder-Traversal/solution_recursive.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def postorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        # recursive
        # tc:O(n); sc:O(h)
        if not root:
            return []
        
        left = self.postorderTraversal(root.left)
        right = self.postorderTraversal(root.right)
                
        left.extend(right)
        left.append(root.val)
        
        return left
        
